### PR TITLE
Audit log phase 1

### PR DESCRIPTION
**Pull request by pedanticantic**
This is phase 1 of the audit log module (obviously).

It is concerned just with creating the audit log records.

It does nothing to report on them - that is phase 2.

### CODE REVIEW

**Comment by pedanticantic** [comment]
I had to update the DB access rules again (because it can create log records for non-logged-in users). It's not perfect, but it's good enough.

This is an extract of the new rules:
`...
    "auditLog": {
      "$auditLog": {
        ".write": "!data.exists() && (auth !== null  || newData.child('userType').val() == 'anonymous')"
      }
    },
    ".read": "true",
    ".write": "auth !== null && !newData.hasChild('auditLog')",
...
`

Basically, by default you can only write records if you are an admin user and you are not trying to write to the audit log. Then, the audit log says, "You can't edit or delete log records at all (the latter may change in phase 3), but you can insert log records if either you are logged in as an admin user, or you are trying to create an "anonymous" log entry".

File: app/scripts/services/DataRepository.js [hunk 1]
```diff
@@ -179,7 +184,8 @@ var DataRepository = function (recordSets) {
     return null;
   };
 
-  this.insert = function insert(collectionName, record, successMsg, callback) {
+  this.insert = function insert(collectionName, record, successMsg, callback, recordId, currentState) {
```
[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/services/DataRepository.js]: You are completely right! I've checked, and they are always the same. I found that I can just use "record" in place of "currentState" - there was no need to create a copy of it, which is nice because it's simpler.

File: app/scripts/services/DataRepositoryFactory.js [hunk 2]
```diff
@@ -6,10 +6,12 @@ var DataRepositoryFactory = function ($firebaseObject, $firebaseArray) {
   this.recordSets = {
     brands: $firebaseArray(db.ref('brands')),
     categories: $firebaseArray(db.ref('categories')),
-    products: $firebaseArray(db.ref('products'))
+    products: $firebaseArray(db.ref('products')),
+    auditLog: $firebaseArray(db.ref('auditLog'))
```
[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/services/DataRepositoryFactory.js]: Thank you, I appreciate that :-)

File: app/scripts/services/DataRepositoryFactory.js [hunk 3]
```diff
@@ -22,7 +24,7 @@ var DataRepositoryFactory = function ($firebaseObject, $firebaseArray) {
     var self = this;
 
     this.syncObject.$loaded().then(function () {
-      self.dataRepository = new DataRepository(self.recordSets);
+      self.dataRepository = new DataRepository(self.recordSets, new AuditLogger(self.recordSets));
```
[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/services/DataRepositoryFactory.js]: I thought about what you were saying, and came to the conclusion that you were right!
When I originally wrote the code I had to decide what was responsible for deciding which audit logger to use (in theory there could be many), and I decided that the DataRepositoryFactory should decide, which is why I did it the way I did (or maybe I just did it like that because the DataRepositoryFactory had the recordSets? I have such a terrible memory!). However, after what you said, I realised we can't pass any old audit logger in, because it has to be consistent with the records that the DataRepository is dealing with (I hope that makes sense), and that means the DataRepository should actually instantiate the audit logger - so in fact, I've gone further than what you suggested, and now it doesn't pass in an audit logger at all, but instantiates it on line 14.
Of course it just happens that the audit logger as it's written doesn't know (and doesn't need to know) anything about the structure of the order records (brands, categories and products), but the point is that the DataRepository needs to make the decision as to which logger to use. Well, I _think_ that's right anyway!!

File: app/scripts/auditLog/models/auditLog.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+var AuditLog = (function () {
+
+  var AuditLog = function AuditLog(data) {
+    this.loggerVersion = 1;
+    this.userType = data.hasOwnProperty('userType') ? data.userType : null;
+    this.userRef = data.hasOwnProperty('userRef') ? data.userRef : null;
+    this.operationType = data.hasOwnProperty('operationType') ? data.operationType : null;
+    this.dateTime = new Date();
```
[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/auditLog/models/auditLog.js]: This is the date/time of an individual log entry, so it's the time of the change (creation, modification, deletion) of the record (brand, category, product).
Picking a log entry at random, it's the "dateTimeUtc" [here](https://console.firebase.google.com/project/end-slavery-now-test/database/end-slavery-now-test/data/auditLog/-Lnme0x8B1W_zeEATAUN).
I don't think "instantiatedAt" is an appropriate name, but we could change it to "eventDateTime" or maybe "createdAt"?
Do you think the name in the database ("dateTimeUtc") needs to change, too?

File: app/scripts/auditLog/models/auditLog.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    var auditLogHelper = new AuditLogHelper();
+
+    // Validate the entries.
+    if (!auditLogHelper.isValidUserType(this.userType)) {
+      // @TODO: Throw some kind of error.
+    }
+    if (!auditLogHelper.isValidOperationType(this.operationType)) {
+      // @TODO: Throw some kind of error.
+    }
+    if (this.delta === null) {
```
[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/auditLog/models/auditLog.js]: It's not right in front you - I've had a look, and it's not initialised and it's the only reference to "delta" in the whole application! You can rest assured you're not going mad!
I think what happened is that I wrote the code for the AuditLog class before I wrote the code that instantiated it, and I was pre-empting the case where an "empty" change was created (ie were a record hadn't changed, but the auditing was invoked for it). When I wrote the code that instantiated an AuditLog, I check first that there is actually a change, and if there isn't, it just aborts, so there is no need for this test, so I've removed it.

[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/auditLog/models/auditLog.js]: Ah, looking at it again, I think it was supposed to be checking "this.changedValues" not "this.delta", so I've reinstated and corrected it. Of course it still does nothing if there are no changes, but we can add something in the future as necessary.

File: app/scripts/auditLog/models/auditLog.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (toriejw) [reply app/scripts/auditLog/models/auditLog.js]: Ahhh my bad! I misunderstood where it was used :) . Makes sense now! I like `eventDateTime` or `createdAt` - those both seem great to me. When I see the field on a record in firebase `dateTimeUtc` makes perfect sense, so I don't think that needs to change (also I have never seen the timezone referenced in the field name, but I love it! the number times I've had to dig to find out "what time zone is this stored in?" .... 🙃)

[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/auditLog/models/auditLog.js]: That's not a problem :-)
I'm glad you like what I did with the timezone. So often I see someone create a database column or a class property and call it something like "height" or "waitTime", and I'm thinking, yes, I can see it's an int/float/whatever but what are the units?? Is it metres, inches, light years, milliseconds, days, what?? I have no idea without trawling through all the code, which is an absolute pain, like you say. Grrr!!! So I always make a point of putting the units, or as in this case, the timezone, in the name of the column/property. So I'd call them "HeightMetres" and "waitTimeMs".
We actually have createdAt and updatedAt on all our tables, and they're set automatically by the database.
I would have been happy with either `eventDateTime` or `createdAt`, but if you're happy with `dateTimeUtc`, I'd prefer to leave it as that as there are now records in firebase with that property name.

[the remark this answers is not in the document]
Comment (pedanticantic) [reply app/scripts/auditLog/models/auditLog.js]: Incidentally, the `loggerVersion` in the [audit log](https://github.com/elixirlabsinc/endslaverynow/blob/audit-log-phase-1/app/scripts/auditLog/models/auditLog.js#L6) model is there so that if anything changes, eg properties are added, old ones are removed or renamed, or datatypes change, we increment the version number with the release that those changes in, and the code that loads the records can process them properly according to the version number.
I've done it in one of my personal app, where it [loads a saved game](https://gitlab.com/pedanticantic/codebreaker/blob/master/src/assets/models/game.ts#L34) (you should have access). It basically takes what it's loaded, and systematically upgrades it to the current version. You'll notice that in the "upgrade to version 1", it introduces the version number (line 85)!
